### PR TITLE
Intellisense broken inside of methods that have delegates as arguments

### PR DESCRIPTION
**Pull request by ivanbasov**
Fix https://github.com/dotnet/roslyn/issues/35027

### CODE REVIEW

**Comment by ivanbasov** [comment]
@jinujoseph please let me know if you think we should fix it for preview 2. Then, we can re-open https://dev.azure.com/devdiv/DevDiv/_workitems/edit/847803 and fill an ask mode template there.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 1]
```diff
@@ -4894,6 +4894,41 @@ namespace ThenIncludeIntellisenseBug
             End Using
         End Function
 
+        <MemberData(NameOf(AllCompletionImplementations))>
+        <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestCompletionInsideMethodsWithDelegatesAsArguments(completionImplementation As CompletionImplementation) As Task
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: Why is this a command handler test versus one in https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs? (I recognize the other tests got added here, but the same question should have been applied to them as well.)

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 2]
```diff
@@ -108,8 +107,7 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
                 {
                     if (method.Parameters.Length > ordinalInInvocation)
                     {
-                        var methodParameterSymbool = method.Parameters[ordinalInInvocation];
-                        var type = methodParameterSymbool.Type;
+                        var type = method.Parameters[ordinalInInvocation].Type;
                         // If type is <see cref="Expression{TDelegate}"/>, ignore <see cref="Expression"/> and use TDelegate.
                         if (type is INamedTypeSymbol expressionSymbolNamedTypeCandidate &&
                             Equals(expressionSymbolNamedTypeCandidate.OriginalDefinition, expressionSymbol))
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Why not the .Equals instance form? If that's binding to object.Equals it's kinda strange to read.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 3]
```diff
@@ -123,13 +121,10 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
...
                         }
 
-                        // If type is <see cref="Func{TSomeArguments, TResult}"/>, extract TSomeArguments 
-                        // and find the type of thee argument corresponding to the ordinal.
-                        if (type is INamedTypeSymbol funcSymbolNamedTypeCandidate &&
-                            Equals(funcSymbolNamedTypeCandidate.Name, funcSymbol.Name) &&
-                            Equals(funcSymbolNamedTypeCandidate.ContainingNamespace, funcSymbol.ContainingNamespace))
+                        if (type.IsDelegateType())
                         {
                             var allTypeArguments = type.GetAllTypeArguments();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: What if the delegate type isn't one with type arguments? Doesn't this have to go looking to the Invoke method or something to get the types of the arguments that way?

File: src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb [hunk 4]
```diff
@@ -3368,6 +3368,33 @@ End Namespace
             End Using
         End Function
 
+        <MemberData(NameOf(AllCompletionImplementations))>
+        <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestCompletionInsideMethodsWithDelegatesAsArguments(completionImplementation As CompletionImplementation) As Task
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb]: We should also have a test with a delegate that isn't a generic delegate, i.e. explicitly have a delegate that hard-codes the type.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 5]
```diff
@@ -92,10 +92,9 @@ protected ImmutableArray<ISymbol> GetSymbols(IParameterSymbol parameter, int pos
         private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candidateSymbols, int ordinalInInvocation, int ordinalInLambda)
         {
             var expressionSymbol = _context.SemanticModel.Compilation.GetTypeByMetadataName(typeof(Expression<>).FullName);
-            var funcSymbol = _context.SemanticModel.Compilation.GetTypeByMetadataName(typeof(Func<>).FullName);
             // We cannot help if semantic model is corrupted or incomplete (e.g. during solution loading) and those symbols are not loaded. 
             // Just bail out.
-            if (expressionSymbol == null || funcSymbol == null)
+            if (expressionSymbol == null)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: I would assume we would skip the special case around Expression<> but the regular delegate handling should work just fine. I'd be more worried if there's some framework out there people target that doesn't have expression tree support.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs (hunk 3 above)
Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: @jasonmalinowski ,  I have added tests for non-generics. Two non-generics work fine (VB and C#). However, one non-generic and another generic do not work together. Semantic models return just a single candidate in those cases. And those candidate even differ in C# and VB. See https://github.com/dotnet/roslyn/issues/35100 for details.

Please let me know if you mean some other tests.

File: src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb (hunk 4 above)
Comment (ivanbasov) [reply src/EditorFeatures/Test2/IntelliSense/VisualBasicCompletionCommandHandlerTests.vb]: I hope I added tests you mean. Please let me know if you see more tests to add.

---
In reply to: [276405342](https://github.com/dotnet/roslyn/pull/35067#discussion_r276405342) [](ancestors = 276405342)

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs (hunk 5 above)
Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Do you see more issues here?

---
In reply to: [276405774](https://github.com/dotnet/roslyn/pull/35067#discussion_r276405774) [](ancestors = 276405774)

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: My worry is we're throwing out the entire behavior change if we can't find Expression, which I could imagine happens in some older portable frameworks or something. I'd just say remove the check here, and add a `expressionSymbol != null` check in your if block after the `If type is <see cref="Expression{TDelegate}"/>, ignore <see cref="Expression"/> and use TDelegate.` comment.

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb [hunk 6]
```diff
@@ -7824,5 +7824,261 @@ End Module
...
+            Await VerifyItemExistsAsync(source, "Task")
+            Await VerifyItemExistsAsync(source, "FirstOrDefault")
+        End Function
+
+        <Fact, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function ThenIncludeSecondArgumentAndMultiArgumentLambda() As Task
+            Dim source = CreateThenIncludeTestCode("0, Function(a, b, c) c.$$",
+"       <System.Runtime.CompilerServices.Extension>
+        Public Function ThenInclude(Of TEntity, TPreviousProperty, TProperty)(
+                                                                             source As IIncludableQueryable(Of TEntity, ICollection(Of TPreviousProperty)),
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb]: Indenting doesn't look right here.

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: My concern isn't the methods being generic, but non-generic delegate types, like:

```
    public delegate void Delegate1(Uri u);
    public delegate void Delegate2(Guid g);

    class Program
    {
        public void M(Delegate1 d) { }
        public void M(Delegate2 d) { }

        public void Test()
        {
            M(d => d.)
        }
    }
```

In this case you should be merging the intellisense for Uri and Guid, correct?

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs (hunk 5 above)
Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Cool! Great idea! Thank you!

File: src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb (hunk 6 above)
Comment (ivanbasov) [reply src/EditorFeatures/VisualBasicTest/Completion/CompletionProviders/SymbolCompletionProviderTests.vb]: It seems to be the standard VB formatting. But I agree it looks weird. Let us move parameters to the left for readability.

---
In reply to: [276726867](https://github.com/dotnet/roslyn/pull/35067#discussion_r276726867) [](ancestors = 276726867)

**Comment by ivanbasov** [comment]
Thank you, @jasonmalinowski ! i added the test you provided and provided its support in code. Please review,

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 7]
```diff
@@ -123,19 +117,38 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
...
+                                var parameters = methods[0].GetParameters();
+                                if (parameters.Length < ordinalInLambda)
+                                {
+                                    continue;
+                                }
+
+                                type = parameters[ordinalInLambda].Type;
                             }
+                            // Functions, Actions and other delegates with TypeArguments.
+                            else
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: You don't need this else branch at all: you should *only* be looking at the Invoke method. Consider the case too like:

`delegate void ReversingArguments<T1, T2>(T2 t2, T1 t1)`

The Invoke method will correctly have the order; the type argument order isn't right.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Or just something like [`EventHandler<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.eventhandler-1) where the generic argument isn't the first argument.

Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Thank you, @jasonmalinowski ! Added a test for this, and moved all the code to use Invoke.

---
In reply to: [276816954](https://github.com/dotnet/roslyn/pull/35067#discussion_r276816954) [](ancestors = 276816954)

Comment (ivanbasov) [reply src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]: Makes sense. Thank you! I think that the reversing argument test covers the switch in code.

---
In reply to: [276817775](https://github.com/dotnet/roslyn/pull/35067#discussion_r276817775) [](ancestors = 276817775)

File: src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs [hunk 8]
```diff
@@ -108,11 +101,12 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
                 {
                     if (method.Parameters.Length > ordinalInInvocation)
                     {
-                        var methodParameterSymbool = method.Parameters[ordinalInInvocation];
-                        var type = methodParameterSymbool.Type;
+                        var type = method.Parameters[ordinalInInvocation].Type;
                         // If type is <see cref="Expression{TDelegate}"/>, ignore <see cref="Expression"/> and use TDelegate.
-                        if (type is INamedTypeSymbol expressionSymbolNamedTypeCandidate &&
-                            Equals(expressionSymbolNamedTypeCandidate.OriginalDefinition, expressionSymbol))
+                        // Ignore this check if expressionSymbol is null, e.g. semantic model is broken or incomplet or if the framework does not contain <see cref="Expression"/>.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs]:
```suggestion
                        // Ignore this check if expressionSymbol is null, e.g. semantic model is broken or incomplete or if the framework does not contain <see cref="Expression"/>.
```

**Comment by ivanbasov** [comment]
@jinujoseph please approve to merge into master (16.1 P3)

**Comment by davidfowl** [comment]
👏

**Pull request merged**